### ISSUE TITLE
[Paper]: Kimi K2.5: Visual Agentic Intelligence

### ISSUE THREAD

**Issue by kyriekevin** [kyriekevin/Latent-Atelier#16]
### Link / Source

https://arxiv.org/abs/2602.02276

**Comment by kyriekevin** [comment]
```json
{
  "summary": "Kimi K2.5 is an open-source large multimodal agentic model that claims state-of-the-art results across text reasoning, coding, vision/video understanding, and computer-use, driven by a training recipe that tightly couples vision and language throughout pretraining and posttraining. The paper’s core multimodal thesis is that early, low-ratio, constant mixing of vision tokens plus a post-training stack (zero-vision text-only SFT followed by joint text-vision RL) yields bidirectional transfer: text bootstraps visual tool-use and visual RL measurably improves text benchmarks. On the agent side, it introduces Agent Swarm with a Parallel-Agent RL (PARL) setup: a trainable orchestrator learns when/how to spawn frozen subagents and schedule subtasks concurrently, improving BrowseComp/WideSearch while cutting latency (reported up to 4.5×). The report additionally provides several engineering-focused mechanisms (e.g., a decoupled vision-encoder process for efficient multimodal training, token-efficient RL “Toggle,” and explicit “critical steps” accounting) intended to make large-scale agentic training stable and cost-aware.",
  "strategic_review": {
    "A1_Significance": {
      "score": 4.4,
      "reason": "Targets two high-leverage bottlenecks for frontier agents: (i) robust multimodal grounding that doesn’t degrade language/coding, and (ii) inference-time latency/scalability limits of sequential tool-use. Parallel orchestration plus joint vision-language optimization maps to problems the community and industry both treat as near-term blockers for real deployments."
    },
    "A2_Novelty": {
      "score": 4.1,
      "reason": "Less a new primitive than a high-complexity system recipe that appears validated at scale: constant-ratio early fusion ablations, zero-vision SFT as a cold-start for visual tool-use, joint multimodal RL with cross-modal gains, and PARL’s frozen-subagent decoupling with anti-collapse reward shaping. Agent Swarm/PARL is the most conceptually novel piece; the rest are strong integrations/choices that, while individually reminiscent of prior practice, are meaningfully composed and stress-tested."
    },
    "A3_Alignment": {
      "score": 4.6,
      "reason": "Directly aligns with critical-path concerns in current agent roadmaps: reducing end-to-end wall-clock latency, enabling wide/deep search behaviors, and scaling effective context via context sharding (multi-agent) rather than truncation. The described metrics (critical steps) and latency improvements speak to operational constraints teams actually optimize."
    }
  },
  "tactical_review": {
    "B1_Rigor": {
      "score": 3.7,
      "reason": "Strong breadth of evaluation (text, coding, agentic, image/video, computer-use) and some targeted ablations (e.g., vision injection timing/ratios; before/after vision RL). However, a noticeable fraction of comparisons are marked as internal (*) and some key training details (exact data filters, reward weights/schedules) are not fully specified, limiting fully independent causal confirmation."
    },
    "B2_Transferability": {
      "score": 3.3,
      "reason": "Several ideas should transfer (early low-ratio fusion; freezing subagents to simplify credit assignment; critical-path/critical-steps accounting; outcome-based visual RL improving calibration). But practical success likely depends on large-scale MoE capacity, extensive proprietary data/infra, and a mature RL stack (GRMs, verifiers), making replication on smaller/public setups non-trivial and potentially brittle without similar tooling."
    },
    "B3_Recipe_Detail": {
      "score": 3.4,
      "reason": "The paper is unusually informative for a frontier-scale technical report: architecture choices (MoonViT-3D), training stages with token counts/sequence lengths, DEP pipeline description, PARL reward structure, and token-efficient RL heuristic are described clearly. Still, some ‘magic numbers’ remain underspecified for reimplementation (e.g., exact modality mixing ratios used in final runs, PARL \u03bb schedules, RL hyperparameters across domains, dataset construction/filters, GRM rubric instantiations), so an external team would likely need additional guesswork."
    }
  }
}
```

**Comment by kyriekevin** [comment]
## 🗺️ Reading Map

### 🎯 The Bullseye (Start Here)
* **Table 1 (Section 2.1):** The key empirical claim behind their *joint pretraining recipe*—early fusion + lower vision ratio beats later, vision-heavy injection **under a fixed token budget**. If this doesn’t hold, the whole “constant ratio throughout training” motivation weakens.
* **Figure 2 (Section 2.2/2.3):** The “zero-vision SFT → scale vision RL FLOPs” story in one plot. This is the main evidence that **text-only SFT is enough to “activate” vision tool-use** and that long-running RL drives continued visual gains.
* **Table 2 (Section 2.3):** The cross-modal-transfer headline: **vision RL improves text benchmarks** (MMLU-Pro, GPQA-Diamond, LongBench v2). This supports their “bidirectional enhancement” narrative (vision refines text, not trade it off).
* **Figure 3 (Section 3):** The actual *Agent Swarm mechanism*—trainable orchestrator + **frozen** subagents, with subagent trajectories excluded from optimization. This is their core engineering choice to avoid credit assignment/instability.
* **PARL reward + CriticalSteps definition (Section 3):** The paper’s “how it learns parallelism” in two pieces:
    * The PARL reward $r_{\mathrm{PARL}}(x,y)$ (instantiation reward + finish-rate reward + outcome reward, with $\lambda_1,\lambda_2$ annealed to 0).
    * The **CriticalSteps** metric $\sum_{t=1}^{T}\left(S_{\mathrm{main}}^{(t)}+\max_i S_{\mathrm{sub},i}^{(t)}\right)$, which explains what they mean by latency/parallel efficiency and what the policy is incentivized to optimize.

---

### ✂️ The Shortcuts (Ignore)
* Skip **Section 1 Introduction** (mostly positioning + claims you’ll validate later).
* Skim **Section 4.1 Foundation: Kimi K2 Base Model** (it’s a recap; only note the MoE scale and “activated params” if you care about deployment cost).
* Skim **Section 4.2 Model Architecture** (unless you specifically care about MoonViT-3D details; mechanism-wise you mainly need: native-resolution + packing + 4-frame temporal compression).
* Skim **Section 4.3 Pre-training Pipeline** (read **Table 3** only; the prose is largely pipeline narration).
* Skim **Section 4.4.1 SFT** (standard data synthesis story; the novel part is *zero-vision SFT*, which is already in Section 2.2).
* Skim **Section 4.4.2 RL – Policy Optimization math** (Equation (1) is implementation detail unless you’re reproducing; focus instead on: what rewards they use + what “Toggle” changes).
* Skim **Section 4.5 Training Infrastructure** (DEP is interesting systems work, but orthogonal to the scientific mechanism claims; read only if you’re optimizing training throughput).
* Skim **Section 5.1.1 benchmark taxonomy + settings** (lists; go straight to **Table 4** if you only need the scoreboard).
* Skip **Section 6 Conclusions** and **References** (no new mechanism).

---

### 🕵️ Delta Check
* [ ] **Check Table 1 carefully:** Did they truly hold the *total* vision+text token budget constant when comparing early/mid/late injection and ratios?  
    * Look for: exact token accounting + whether any other variables changed (sequence length, data mix, optimizer schedule).
* [ ] **Check the “zero-vision SFT works” causality (Figure 2 + surrounding text):**  
    * Is there a clean comparison vs **text-vision SFT** (they claim it’s worse) with the same data quality control and RL compute?  
    * Or is the conclusion mostly attributed to “vision SFT data is low quality” without a controlled alternative?
* [ ] **Check the “vision RL improves text” claim (Table 2):**  
    * Are the “before” vs “after” checkpoints matched for total training compute/steps, or did “after vision-RL” simply get more optimization time overall?  
    * If compute isn’t matched, does the paper provide evidence it’s *specifically* visual RL (not extra training) driving gains?
* [ ] **Check Agent Swarm fairness (Table 6 + Figure 8 + CriticalSteps):**  
    * Are single-agent baselines given the same tool budget / wall-clock / cluster resources?  
    * Does “up to \(4.5\times\)” latency reduction hold under the **CriticalSteps** constraint, and is the metric aligned with real wall-clock in their setup?
